REORG: JoinStyle and CapStyle classes

brunobeltran opened the pull request:
Centralize docs and validation for JoinStyle and CapStyle in one place.

On top of <strike>#18817</strike> #19063.

## PR Summary

First GSOD PR. Follow suggestion of @anntzer in [MEP 30](#17446) to simply start making the simple string enum cases before moving onto more complicated things like LineStyle.

The primary goal of this PR is to create a centralized place to document what is a `JoinStyle` anytime that parameter is used:
```
class Collection(...):
    ...
    def __init__(...):
    """
    ...
        joinstyle : `.JoinStyle`-like, default: :rc:`patch.joinstyle`
            Style to use for joining lines for all paths in the collection.
    """"
```

This PR accomplishes this by adding (totally transparent) `JoinStyle` and `CapStyle` classes, which act as both a centralized location for documentation of these "features" of our library, and as validators (and, hopefully, in the future, serializers for rc).

The user can still use the standard string input style, i.e. this still works:
~~~
plt.plot(x, y, joinstyle='miter')
~~~

Up until this point, I am basically just implementing the GSOD in the way which was agreed upon in the [last dev call about this](https://hackmd.io/MNcskNfZSIOoLb50GqVIGg#6-July-2020). But there are some additional considerations that come up once you start to implement this.

As soon as any end-user-facing function takes a `joinstyle` argument, we have two options. Either we can immediately convert:

~~~
def __init__(..., joinstyle=JoinStyle.round):
... # ValueError if invalid `joinstyle`
    self._joinstyle = JoinStyle(joinstyle)
~~~

or we can validate the input but store the string:

```
def __init__(..., joinstyle='round'):
... # ValueError if invalid `joinstyle`
    self._joinstyle = joinstyle if JoinStyle(joinstyle)
```

The former has the upside of being in line with what we already do elsewhere in the library (e.g. `MarkerStyle`), while the latter has the upside of breaking less API surface naively (EDIT: although, as I point out to Tim in the comments below, this can be very strongly mitigated with enough care in implementing dunder methods for backcompat).

This PR implements the approach where we immediately convert (the former of the two above). 


## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

(Note: Tom explicitly does not want this new API surface advertised (yet)).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

brunobeltran commented:
Alright @timhoffm, these docs now look like what I would expect when I build them locally. Feel free to offer any feedback, but some points I'm most interested in your opinion are:

1) How to deal with "constructible-as". I made sure that everywhere which now takes a `[Cap/Join]Style` still takes the old string inputs by simply calling the new constructors immediately. This means that functions with those arguments will have `Parameters` entries which look like:
```
joinstyle : `JoinStyle` or {'bevel', 'miter', 'round'}, default: :rc:`patches.joinstyle`
```
or similar. 

Frankly, this is totally fine, but for other types, like `LineStyle`, this will quickly get out of hand. I propose the following standard:
we list out both the "`_types`" type *and* the alternative options (e.g. `LineWidth or float`) *unless* that would make us overflow the line. This has the upside of keeping things as easy to discover as possible without sacrificing readability (and I would argue that if you don't know what all the options for e.g. `LineStyle` are, I do want to force you to click that link or type "help()" once more.

Another option is to do something similar to what `numpy` does, and have some syntax for "this parameter can be any type that can be passed directly to the constructor of X". "X-like", "X-able", "X-constructible" or something similar. This would make the above docstring look like
```
joinstyle : `JoinStyle`-like, default: :rc:`patches.joinstyle`
```
which is of course much shorter. 

2) Obviously I'm a fan of your suggestion to call this module `_types.py`, but any other naming thoughts are more than welcome!

3) and finally, I guess do we want this to get "what's new"? I guess if we want this to be solely a documentation/internal validator mechanism for now, I could just make sure the entry emphasizes that this private module is open for use, but that it's name/etc. are liable to change in a future version?

brunobeltran commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+from matplotlib import cbook
+
+
+def _deprecate_case_insensitive_join_cap(s):
+    s_low = s.lower()
+    if s != s_low:
+        if s_low in ['miter', 'round', 'bevel']:
+            cbook.warn_deprecated(
+                "3.3", message="Case-insensitive capstyles are deprecated "
+                "since %(since)s and support for them will be removed "
+                "%(removal)s; please pass them in lowercase.")
+        elif s_low in ['butt', 'round', 'projecting']:
+            cbook.warn_deprecated(
+                "3.3", message="Case-insensitive joinstyles are deprecated "
+                "since %(since)s and support for them will be removed "
+                "%(removal)s; please pass them in lowercase.")
+        # Else, error out at the check_in_list stage.
+    return s_low
```
Deprecation machinery copied as-is out of `rcsetup.py`.

jklymak replied:
This seems arbitrary, and should definitely be a class method if you really want to keep it in.  Why would we care if join styles are capitalized or not?  Do we really think "round" and "Round" will ever mean something different?

brunobeltran replied:
This is existing behavior. I was not around for the discussion about whether or not to allow caps. I agree that for a user-facing thing, allowing case-insensitivity is probably fine, but this should probably not be hashed out in this PR.

brunobeltran commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+        Enum.__init__(self)
+
+    @staticmethod
+    def demo():
+        import numpy as np
+        import matplotlib.pyplot as plt
+
+        def plot_angle(ax, x, y, angle, style):
+            phi = np.radians(angle)
+            xx = [x + .5, x, x + .5*np.cos(phi)]
+            yy = [y, y, y + .5*np.sin(phi)]
+            ax.plot(xx, yy, lw=12, color='tab:blue', solid_joinstyle=style)
+            ax.plot(xx, yy, lw=1, color='black')
+            ax.plot(xx[1], yy[1], 'o', color='tab:red', markersize=3)
+
+        fig, ax = plt.subplots(figsize=(8, 6))
+        ax.set_title('Join style')
+        for x, style in enumerate(['miter', 'round', 'bevel']):
+            ax.text(x, 5, style)
+            for y, angle in enumerate([20, 45, 60, 90, 120]):
+                plot_angle(ax, x, y, angle, style)
+                if x == 0:
+                    ax.text(-1.3, y, f'{angle} degrees')
+        ax.set_xlim(-1.5, 2.75)
+        ax.set_ylim(-.5, 5.5)
+        ax.set_axis_off()
+
+        plt.show()
```
Demo copied as-is from `gallery/lines_bars_and_markers/joinstyle.html`

timhoffm replied:
I'm not quite sure about the demo. On the one hand, it may be nice to be able to display the options. On the other hand, I don't like the code duplication with the example. That will get out of sync sooner or later. Also, if the user is calling `demo()` while working on a figure, `plt.show()` will display that figure and thus mess up the user's work.

brunobeltran replied:
Ah yes, sorry, I meant to remove that example, it's gone now.

Obviously I don't think this type of documentation should *replace* examples, but in this case (where the example was literally just a demo of the definitions of each available style), I think that particular documentation should live with the class, where it won't go out of sync.

> Also, if the user is calling demo() while working on a figure, plt.show() will display that figure and thus mess up the user's work.

This is a good point, I changed the `demo` to only "show" its own figure.

brunobeltran commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+
+    @staticmethod
+    def demo():
+        import matplotlib.pyplot as plt
+
+        fig, ax = plt.subplots(figsize=(8, 2))
+        ax.set_title('Cap style')
+
+        for x, style in enumerate(['butt', 'round', 'projecting']):
+            ax.text(x+0.25, 1, style, ha='center')
+            xx = [x, x+0.5]
+            yy = [0, 0]
+            ax.plot(xx, yy, lw=12, color='tab:blue', solid_capstyle=style)
+            ax.plot(xx, yy, lw=1, color='black')
+            ax.plot(xx, yy, 'o', color='tab:red', markersize=3)
+        ax.text(2.25, 0.7, '(default)', ha='center')
+
+        ax.set_ylim(-.5, 1.5)
+        ax.set_axis_off()
```
Demo copied as-is from `gallery/lines_bars_and_markers/joinstyle.html`

brunobeltran commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -24,16 +24,17 @@
+from matplotlib._types import JoinStyle, CapStyle
 from matplotlib.texmanager import TexManager
 from matplotlib.transforms import Affine2D
-from matplotlib.backends.backend_mixed import MixedModeRenderer
```
alphabetization

brunobeltran commented on lib/matplotlib/collections.py:
```diff
@@ -110,14 +111,10 @@ def __init__(self,
             in points. For examples, see
             :doc:`/gallery/lines_bars_and_markers/linestyles`.
-        capstyle : str, default: :rc:`patch.capstyle`
+        capstyle : `.CapStyle`-like, default: :rc:`patch.capstyle`
```
@timhoffm This would be an example where including the individual possibilities would push us over a  single line.

brunobeltran commented on lib/matplotlib/rcsetup.py:
```diff
@@ -22,10 +22,11 @@
 
-from matplotlib import _api, animation, cbook
+from matplotlib import animation, cbook
 from matplotlib.cbook import ls_mapper
-from matplotlib.fontconfig_pattern import parse_fontconfig_pattern
 from matplotlib.colors import is_color_like
+from matplotlib.fontconfig_pattern import parse_fontconfig_pattern
+from matplotlib._types import JoinStyle, CapStyle
```
tl;dr: alphabetized, added _types, removed _api

brunobeltran commented:
CI failing on backends that I don't have installed locally (surprise). I'll do a better job of cleaning that up ASAP.

But this also goes to show that with this specific proposal, this is a major API change for people who write their own backends. I'll bring this up on the next call.

timhoffm commented:
## tl/dr;
Not yet decided. I have to think about this more.


### Private/public

I'm hesitant to make this public right now. Using these types on the front-end is a lot bulkier:
~~~
from matplotlib._types import JoinStyle
plt.plot(x, y, joinstyle=JoinStyle.miter)
~~~
instead of 
~~~
plt.plot(x, y, joinstyle='miter')
~~~
I don't feel like recommending the above.


### Type documentation

In turn, if `JoinStyle` is not public, we should not announce it as a supported parameter type.

#### For completeness: If the types were public:

1. We should try not to overflow the line.
2. Explicit is better:
   If it matches on the line:
   ~~~
   joinstyle : `JoinStyle` or {'bevel', 'miter', 'round'}
   ~~~
   otherwise (we already use "array-like"):
   ~~~
   joinstyle : `JoinStyle`-like
   ~~~

#### Alternatives

If you are motivated to jump into sphinx directives:
~~~
joinstyle : :mpltype:`JoinStyle`
~~~
With a suitable rendered representation. The downside is that inline help does not have the explicit values.

story645 commented:
There is the "provisional" label for making it semi-public if you want a middle ground.

brunobeltran commented:
> I'm hesitant to make this public right now. Using these types on the front-end is a lot bulkier:

The way the type is constructed is completely transparent to the user. I would argue that even if this was fully implemented, we would still use the "string" versions in all of our examples, and encourage users to do so.

in other words, this still works:

> ```
> plt.plot(x, y, joinstyle='miter')
> ```
IMO the real issue isn't whether or not we promote this to our *end* users (since they are really not affected except for they get the extra option of passing in these guys as the explicit object if they *really want*), but whether we internally store the string (e.g. in `GraphicsContextBase`) or store the new Enum object (in which case third-party backends would have to add `.name` anywhere they use the `JoinStyle` as a dictionary key)...[see footnote]

> I can't bring myself to recommending or even only showing the above.
The benefits are clearly not there for a simple case like `JoinStyle`, but the goal is to make objects like `LineStyle` more useable in addition to making their advanced features more discoverable (similarly to how `MarkerStyle` gets its own class, but nobody complains because you can still just pass 'o' or similar).


> If you are motivated to dive into sphinx roles:
> ```
> joinstyle : :mpltype:`JoinStyle`
> ```
This was an alternate proposal that I fully wrote up at one point. In the last dev call where this was discussed, it was decided that making them full classes makes more sense, as this gives us a place to put validators/serializers (since I'll be writing the validators anyway given that it's zero extra work once I commit to fully documenting these things).

@timhoffm if you think it would be worthwhile I'm happy to have this role implemented for Monday, since I've done the sphinx role dance before, if you think having these two options would aid discussion?

....

[footnote]
...although as a side note, we could even do an arbitrarily good job handling backcompat, by overwriting `__hash__` and changing the default `Enum` dunder functions to be more friendly, e.g.:
```
def __eq__(self, other):
    try:
        other = self.__class__(other)
    except ValueError:
        return False
    return self == other
```

and so forth. I could make a "Wasn'tAClassBeforeBase" class to share all of these "fixes"....

jklymak commented:
I think this is useful, so long as the old behaviour is maintained of allowing a string.  The point of making these public is that they will show up as their own thing in the docs.  

OTOH, I don't know about calling this module "types".  That is a pretty generic and loaded term in computing.  `style_classes`?  Dunno, but not `types`!

timhoffm commented:
@jklymak We already have a [style](file:///home/tim/dev/matplotlib/doc/build/html/api/style_api.html) module. That name is taken and would be too confusing to be used with a different context.

## Requirements
(sort of, style classes are already a solution, but hey)
- we want the style classes for internal stuff (validating, serializing, maybe as internal datatype)
- we want a dedicated place to put docs for these and have links from corresponding parameters.
- user should still only use the simple types (e.g. strings)

## Proposed solution

1. Let's keep the style classes private. If users shouldn't use them, they don't need to know that there is a formal `JoinStyle` class.
2. Style classes collect all the desired functionality, including documentation.
3. We don't put the classes in our public docs using the standard sphinx API generation tools (these classes are private, see 1.). Instead, we need another mechanism to extract the documentation and put it on some pages, possibly a custom sphinx directive. To be decided whether these docs are the class docstring or some other string attached to the class.
4. We also need a mechanism to link types, possibly a sphinx role. Maybe (up for discussion):
  ``:mpltype:`joinstyle` `` with an option for ``:mpltype`joinstyle {'miter', 'round', 'bevel'}` `` if we want to state explicit values for plaintext docs.

Defining own sphinx roles and directives is a bit more heavy-weight, but it allows us to use internal code structures and keep them completely hidden from the user.

  ~~~

jklymak commented:
But a counterpoint is that if we have public classes they could take kwargs on init.  I imagine some of those have parameters that could be adjusted.  We do exactly this for markers.

timhoffm commented:
(Some of) the classes can still be made public if we feel that is beneficial.

jklymak commented:
I'd agree with you, if you hadn't just proposed a new sphinx directive and a new sphinx role.  The more of these we have, the more fragile our development process is.

brunobeltran commented:
>     * user should still only use the simple types (e.g. strings)

@timhoffm, this phrasing is surprising to me, based on the concerns you voiced above, I would have expected you to say:

> *user should still be able to use the the simple types (e.g. strings)

I agree, and I think no proposal can move forward that doesn't allow the user to still use the string syntax, but I don't see the danger/negative to allowing the user to use this new syntax....

In particular, this would solve the tab-complete dilemma that I've seen you be actively involved in before, right? I mean if the user does take the time to `from _types import *`, then they can do `..., lw=5, joinstyle=JoinStyle.[TAB]`.

And as we all know, documentation-by-available-tab-completion-options is a real thing.

brunobeltran commented:
> That is a pretty generic and loaded term in computing. `style_classes`? Dunno, but not `types`!

I had originally called the module `_styles`, but was 'suaded by Tim's argument about it being too close to "style". I would argue that the whole module is private for now so we are free to hash out synonyms to "styles" or "types" for some time before we commit.

timhoffm commented:
> I don't see the danger/negative to allowing the user to use the new syntax, even if in this case it's strictly more verbose....

On a technical level that should work. I'm more concerned on "there should only be one way" for the parameters. Are the new classes hidden / mentioned but not emphasized / an equal alternative? For now, I tend towards not mentioning them to the user at all to not increase the perceived API complexity.

You have a point with tab-completion. Unfortunately, jedi will indeed not support string completion from docstrings. However, there's a chance to get string completion from typing at some point https://github.com/davidhalter/jedi/issues/1355.

timhoffm commented:
> I'd agree with you, if you hadn't just proposed a new sphinx directive and a new sphinx role. The more of these we have, the more fragile our development process is.

I disagree on fragile. If they are well written and documented, additional roles do not complicate development. In contrast, they can help to generate better, more consistent documentation. E.g. you could auto-generate the possible string values in the HTML docs. Or, if you want explicit values for text docstrings like :mpltype`joinstyle {'miter', 'round', 'bevel'}`, the role could validate that the stated texts are correct.

brunobeltran commented:
> On a technical level that should work. I'm more concerned on "there should only be one way" for the parameters. 

Hopefully not getting too philosophical here, but while I am always behind have "One Way™" to do things, I don't think allowing the user more options for the *type* to be passed as a parameter counts as having "more ways". To the contrary, I think it's pretty standard fare---especially in the scientific Python ecosystem---to expect that the input you pass to a function might be converted internally to a more precise representation....I mean even setting numpy aside look at dict_keys or other things even in core Python. From my perspective the etiquette for doing something like this as a library writer seems to be that it's "Okay™" as long as **the user doesn't have to know about the internal representation to get the job done** (which is still true here).

so I don't know if any users would even blink about it any more than they would be confused whether they are allowed to pass a list or a tuple to certain inputs....

...okay fine so users do get confused about that but I would argue that's just how things are done in Python.

> Are the new classes hidden / mentioned but not emphasized / an equal alternative? For now, I tend towards not mentioning them to the user at all to not increase the perceived API complexity.

I don't see any reason reason to mention them, but presumably they will be discovered naturally by users that are reading the documentation deeply anyway, so I don't see any reason to "hide" them either. As long as both the string and enum versions still quack, I don't see users getting upset.  

> You have a point with tab-completion. Unfortunately, jedi will indeed not support string completion from docstrings. However, there's a chance to get string completion from typing at some point [davidhalter/jedi#1355](https://github.com/davidhalter/jedi/issues/1355).
How this interacts with Typing is an interesting, but I guess very future discussion, (but IMO having it be a "real" class would greatly simplify any implementation of "types" in the future, even thought we keep the "string" versions forever).

jklymak commented:
Again, we do this for markers; either take a string or a class, and I don't think its a problem at all, and leaves the door open for users to write new styles.  We could do the same for line styles, and that would get us part way towards allowing folks to write their own brush stroke patterns for lines, or complicated dot-dash patterns without ugly strings to parse.

timhoffm commented:
One of the questions influencing the solution space is how we want the docs in rendered and source from to describe these types.

Possible rendered docs:
<div>
1) joinstyle : {'miter', 'round', 'bevel'}
</div>
<div>
2) joinstyle : <a href="">JoinStyle</a>
</div>
<div>
3) joinstyle : <a href="">JoinStyle</a>-like
</div>
<div>
4) joinstyle : <a href="">{'miter', 'round', 'bevel'}</a>
</div>
<div>
5) joinstyle : <a href="">JoinStyle</a> {'miter', 'round', 'bevel'}
</div>

Possible raw docs:
~~~
1) joinstyle : {'miter', 'round', 'bevel'}

2) joinstyle : `JoinStyle`

3) joinstyle : `JoinStyle`-like

4) joinstyle : :mpltype:`{'miter', 'round', 'bevel'}`

5) joinstyle : :mpltype:`JoinStyle {'miter', 'round', 'bevel'}`

6) joinstyle : :mpltype:`JoinStyle`
~~~

1) is the current way and not what we want as it has no link.
2)+3) Could be used with type classes and standard numpydoc/sphinx documentation machinery.
4)+5)+6) Need a custom sphinx role. Due to each of the raw docs could generate either rendered version.
Raw 4) needs a special internal lookup, that associates "{'miter', 'round', 'bevel'}" with `JoinStyle`.

I'm inclined towards 4). Because it states the string values most users would use both in raw and HTML, but still links to the style docs.

jklymak commented:
How about:
```python
joinstyle : {'miter', 'round', 'bevel'} or `JoinStyle`
```

timhoffm commented:
Yes added to the list and scheduled for discussion in the dev call.

brunobeltran commented:
There are two discussions happening here now.
1. Whether or not we should make a class for JoinStyle, CapStyle, LineStyle, Hatch, Color, URL, ...., LineWidth?, etc.
2. How to render the documentation, which I claim is largely independent (except for that obviously if there is no class, we have to use a Sphinx directive somewhere).

For (1), I'd vote a strong yes, and I wager I can convince the team on the call, as I've now done so once before already ;)

How to render this class, on the other hand, will definitely require more input. All of the options you list seem to work well for the (pretty trivial) case of a `JoinStyle`, but in general, we also include the "default" value in any given docstring:

```
joinstyle : `.JoinStyle`-like, default: :rc:`patch.joinstyle`
```

and with this addition, almost all of the options listed above would overflow 80 characters, even for `JoinStyle`, which does not even really have a very verbose list of possible inputs. Looking forward a bit, (e.g. for `LineStyle`), we'll need to have a more robust system, since even the possible string inputs by themselves don't fit on a line as-is. I'd wager that instead of considering all the possible permutations listed above, it may be more efficient to get consensus on the following basic questions:

1. What to call the "real" class (if anything at all):
    a. Nothing
    b. `SomeStyle`
    c. `SomeStyle`-like
    d. `SomeStyle`-constructible
    e. :mpltype:`SomeStyle`
2. How to make space to fit everything on a line:
    a. Move default value into description if needed.
    b. Shorten possible values (e.g. {'miter', ...}) if needed.
    c. Move possible values into description if needed.
    d. Any of the above even when not needed.
3. How to signal that there are more options besides what is listed explicitly:
    a. Explicitly in the description.
    b. By using "-like" when not all options are shown.
    c. Both.

I would advocate for always using the `-like` when all other options aren't listed explicitly, with the whole idea being to convey to the user the idea that anything that "can be constructed into a LineStyle" works. In cases where all options are listed, we should leave it out:

```
joinstyle : {'miter', 'round', 'bevel'} or `JoinStyle`
```

When this format is possible, leading with the concrete values first as @jklymak suggests is also great because it implicitly communicates that we expect most users to just use the string versions, but realistically this will rarely be possible.
With the typical 8 spaces starting the line, then even as super-shortened form overflows the line:

```
joinstyle : {'miter', ...} or `.JoinStyle`, default: :rc:`patch.joinstyle`
```

Which means we will typically have to sacrifice either the defaults or the allowed strings (or both) to the parameter description itself

```
	"""
	linestyle: {'-', 'solid', '--', 'dashed', ...} or `LineStyle`-like
	    Customize the dashing pattern of the line. See the `LineStyle` 
	    docs for all possible inputs. Default: 'solid'.
	"""
```

or, (my preference):


```python
	"""
	linestyle: `LineStyle`-like, default: 'solid'
	    Customize the dashing pattern of the line. Values that can be
	    automatically interpreted as a `LineStyle` include {'-', 'solid',
            '--', 'dashed', ':', 'dotted', '-.', 'dashdot'}.
	"""
```

I like this last version the move by far, because it accomplished a few goals well:

1. The "default" is up-front, so if that's what you're scanning for, you don't have to read the description.
2. `\`SomeStyle\`-like, default: :rc:\`whatever.somethingstyle\`` will always fit in one 80-character line. 
3. The description leads with what the parameter **does**, not just possible options for the parameter, as is enforced by other projects that use numpydoc (this is currently (largely) *not* the case in our docs).
4. The most typically options are all clearly listed, while explicitly hinting that there are other ways to specify linestyle (which would lead interested users to click through, while also not requiring your typically user to click through (or, god forbid, try to search for the LineStyle class docs in the terminal)).

jklymak commented:
I think if the options are too long you just put it on the next line?
```
joinstyle : string or `JoinStyle`
    string is one of {'a', 'b', 'c'} 
```

or if that is too verbose, you say 
```
    where string is one of the strings documented in `JoinStyle`, e.g. 'a', 'b', and 'c'. 
```

brunobeltran commented:
If I'm correct in assuming that would look like the following when fully fleshed out:

> ```
> joinstyle : str or `JoinStyle`, default: :rc:`patch.joinstyle`
>     Style to use when drawing connected line segments. Allowed values are {'round', 'miter', 'bevel'}. 
> ```

then yeah, I believe that only differs from my favorite proposal in that it includes `str or ``JoinStyle```, which I am totally fine with in principle, but the original reason I preferred my spelling was the concern of, how do we decide when to cut off the alternate type specifications? If we do something like:

```
"""
        linestyles : str or tuple or `LineStyle`, default: :rc:`patch.linestyle`
"""
```

then I feel like we're starting to lose the whole point of having the class available for users to click through to the complete parameter description....and once you start to do `color` (which is next on my list, about to push up linestyle), then you *really* lose the whole point if you start to try to spell out all the alternate types.

I like the idea of (for the sake of uniformity), just having `SomeStyle`-like everywhere. If we do `str or ``SomeStyle``` in some places and in other places we mean `str or tuple or list of tuple or ``SomeStyle``` but we only can *say* `str or ``SomeStyle``` for space reasons, then I worry we would confuse users or be falsely suggesting that there's not more advanced ways to do things.

jklymak commented:
Sure, but there are middle grounds.  If you want to say "JoinStyle-like" thats fine, but don't let elegant get in the way of helpful .   If possible I'd suggest that if there are a few options, just lay them out rather than force the user to click.  If its complicated, like color, I'd still suggest giving a couple of examples: "For example, 'r', 'green', 'blue', or RGBA tuples like: [1.0, 0.5, 0.5, 1.0];  for the full list, see `ColorStyle`".

brunobeltran commented:
I fully agree that the 90%-ish use cases should have explicit examples in the docstring itself (not requiring clickthrough), definitely didn't mean to suggest otherwise.

It sounds like the real meat of the issue (i.e. where do we draw the line with which examples we are going to include directly in the description and which will require the user to click through) may have to happen on a case-by-case basis.

QuLogic commented:
You should be able to deprecate `Line2D.validCap` and `Line2D.validJoin` now.

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+
+class CapStyle(Enum):
+    """
+    Define how the the end of a line is drawn.
```
```suggestion
    Define how open end points of a line are drawn.
```

brunobeltran replied:
While I immediately see what you mean, I don't think "open" vs "closed" will be the most enlightening for a user actually using this docstring to figure out what a `CapStyle` **is**.

Is what I have now better?

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+    """
+    Define how the the end of a line is drawn.
+
+    How to draw the start and end points of lines that represent a closed curve
+    (i.e. that end in a `~.path.Path.CLOSEPOLY`) is controlled by the line's
+    `JoinStyle`. For all other lines, how the start and end points are drawn is
+    controlled by the *CapStyle*.
```
```suggestion
    This does not affect the end points of closed curves (i.e. `~.path.Path`\s
    that end in a `~.path.Path.CLOSEPOLY`), which are instaed controlled by
    the `JoinStyle`.
```
Note: You need to make this a raw string then because of `` `~.path.Path`\s``.

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+    `JoinStyle`. For all other lines, how the start and end points are drawn is
+    controlled by the *CapStyle*.
+
+    For a simple visual description of each *CapStyle*, `view these docs
```
```suggestion
    For a visual impression of each *CapStyle*, `view these docs
```

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+    Matplotlib provides three options for defining how the corners where two
+    segments meet should be drawn. In short:
+
+        - *miter* is the "arrow-tip" style. Each boundary of the filled-in area
```
You should not indent bullet points in ReST. That gives additional unwanted indent. Instead this should look like
~~~
    some text:

    - first item foo bar
      more text
    - second item
~~~

Yes, ReST is picky!

brunobeltran replied:
Thanks for the heads up! I'm always impressed by how well everyone here keeps RST rules lined up. I switch so much between .md and .rst that I live in perpetual mix-up land.

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+        from matplotlib._types import CapStyle
+        CapStyle.demo()
+
+    Briefly, the three options available are:
```
```suggestion
    Available options:
```

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+
+    Briefly, the three options available are:
+
+        - *butt*: the line is squared off at its endpoint.
```
No indent for bullet points.

timhoffm commented on lib/matplotlib/backend_bases.py:
```diff
@@ -820,7 +820,7 @@ def get_antialiased(self):
 
     def get_capstyle(self):
         """
-        Return the capstyle as a string in ('butt', 'round', 'projecting').
+        Return the default `.CapStyle`.
         """
```
If the summary line is short enough, we put it on one line, including the quotes:

~~~
    def get_capstyle(self):
        """Return the default `.CapStyle`."""
~~~

timhoffm commented on lib/matplotlib/backend_bases.py:
```diff
@@ -866,7 +866,7 @@ def get_forced_alpha(self):
 
     def get_joinstyle(self):
-        """Return the line join style as one of ('miter', 'round', 'bevel')."""
+        """Return the `.JoinStyle`."""
```
For `get_capstyle` you've added "default". Not quite sure which is best "default", "current", blank or something else. Anyway, it should be consistent for `get_capstyle` and `get_joinstyle`.

brunobeltran replied:
I'll change it back to blank, to be consistent with other properties of the GC.

timhoffm commented on lib/matplotlib/backend_bases.py:
```diff
@@ -919,9 +919,15 @@ def set_antialiased(self, b):
-        validate_capstyle(cs)
-        self._capstyle = cs
+        """
+        Set the `.CapStyle`.
+
+        Parameters
+        ----------
+        cs : `.CapStyle` or {'butt', 'round', 'projecting'}
+            How to draw end points of lines.
```
Not a strong opinion, but this may read a bit better:

```suggestion
        Set how to draw end points of lines.

        Parameters
        ----------
        cs : `.CapStyle` or {'butt', 'round', 'projecting'}
```

timhoffm commented on lib/matplotlib/backend_bases.py:
```diff
@@ -987,9 +993,14 @@ def set_foreground(self, fg, isRGBA=False):
-        validate_joinstyle(js)
-        self._joinstyle = js
+        """
+        Set the `.JoinStyle`.
```
Similarly, 
```suggestion
        Set how to draw the connections of line segments.
```

timhoffm commented on lib/matplotlib/lines.py:
```diff
@@ -1309,83 +1310,83 @@ def set_dash_joinstyle(self, s):
-        s : {'miter', 'round', 'bevel'}
-            For examples see :doc:`/gallery/lines_bars_and_markers/joinstyle`.
+        s : `.JoinStyle` or {'miter', 'round', 'bevel'}
+            How to join segments of the line that are dashed.
```
```suggestion
            How to join segments of dashed lines.
```

brunobeltran replied:
I think I wanted to allude (since a user might not know when reading this) to the fact that the line can be *either* solid or dashed. I fixed the wording to do this more directly.

timhoffm commented on lib/matplotlib/lines.py:
```diff
@@ -1309,83 +1310,83 @@ def set_dash_joinstyle(self, s):
-        s : {'miter', 'round', 'bevel'}
-            For examples see :doc:`/gallery/lines_bars_and_markers/joinstyle`.
+        s : `.JoinStyle` or {'miter', 'round', 'bevel'}
+            How to join segments of the line that are solid.
```
```suggestion
            How to join segments solid lines.
```

timhoffm commented on lib/matplotlib/lines.py:
```diff
@@ -1309,83 +1310,83 @@ def set_dash_joinstyle(self, s):
-        s : {'butt', 'round', 'projecting'}
-            For examples see :doc:`/gallery/lines_bars_and_markers/joinstyle`.
+        s : `.CapStyle` or {'butt', 'round', 'projecting'}
+            How to draw end point when line is dashed.
```
```suggestion
            How to draw end points of dashed lines.
```

timhoffm commented on lib/matplotlib/lines.py:
```diff
@@ -1309,83 +1310,83 @@ def set_dash_joinstyle(self, s):
-        s : {'butt', 'round', 'projecting'}
-            For examples see :doc:`/gallery/lines_bars_and_markers/joinstyle`.
+        s : `.CapStyle` or {'butt', 'round', 'projecting'}
+            How to draw end point when line is solid.
```
```suggestion
            How to draw end points of solid lines.
```

timhoffm commented:
Mostly wording. In a few places you can be shorter while still conveying the same information. People don't like to read docs, so shorter is better :smile:.

brunobeltran commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,165 @@
+        JoinStyle.demo()
+
+    For a more precise description, we first recall that lines in Matplotlib
+    are typically defined by a 1D `~.path.Path` and a finite ``linewidth``,
```
The more correct thing to say that 

    in Matplotlib, any line that will eventually be drawn must have at least three features:
        1. a 1D `~.path.Path` defining the path it traces out
        2. a `~.transforms.Transform` mapping the coordinates of the `~.path.Path` to display space
        3. a finite ``linewidth``

however, I thought that adding the discussion of `Transforms` would simply obfuscate the real issue at hand.

It is *True* that in order to be drawn at the end of the day, a line must have a path and a width though. While I haven't checked *every* instance of a thing that looks "like a line" in our library (there are several besides Line2D, for sure, like `MarkerStyle` outlines)...I think I'd go as saying I would not call anything with at least these two properties a line.

anntzer commented on src/py_converters.cpp:
```diff
@@ -130,7 +130,17 @@ int convert_cap(PyObject *capobj, void *capp)
 
-    if (!convert_string_enum(capobj, "capstyle", names, values, &result)) {
+    PyObject *name = PyUnicode_FromString("name");
+    PyObject *cap_string = PyObject_GetAttr(capobj, name);
```
you can probably use https://docs.python.org/3/c-api/object.html#c.PyObject_GetAttrString instead.

brunobeltran replied:
Will fix in the future, for now, Tom has asked that I revert a large chunk of this to make it so that we don't expose the JoinStyle and CapStyle classes publicly *at all* for now.

brunobeltran commented:
Per the most recent GSOD call, we've decided to not expose the new classes at all to users, even if they try to ask for them.

Users are free to pass `JoinStyle`/`CapStyle` objects in if they please (for some reason), but all classes will emit the string representation on `get_joinstyle`/`get_capstyle`, regardless of what is passed in.

This is to ensure consistency moving forward, as we will be doing a similar thing for LineStyle (returning the user back exactly the format of linestyle that they passed in).

@timhoffm this should address all of your wording concerns.

jklymak commented:
Is it really a problem from a user point of view to pass in a string and get back a class that represents that string in a more feature-full way?  The typical use case of getting the format back is to use it for another artist so returning a full fledged class is fine for that.

timhoffm commented:
Unfortunately, our user-base is large enough that we also have lots of atypical use cases. I don't dare changing the return type for now. It's too large an API break.

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,179 @@
+        ax.set_axis_off()
+        fig.show()
+JoinStyle.input_description = "{" \
+        + ", ".join([js.name for js in JoinStyle]) \
```
Quote every single name. These are strings and should look like ones.

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,179 @@
+        ax.set_axis_off()
+        fig.show()
+CapStyle.input_description = "{" \
+        + ", ".join([cs.name for cs in CapStyle]) \
```
Quote every single name. These are strings and should look like ones.

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,179 @@
+        + "}"
+
+
+class CapStyle(str, _AutoStringNameEnum):
```
I'm not particularly happy with the overall appearance of the docs (https://48026-1385122-gh.circle-artifacts.com/0/doc/build/html/api/_types.html#matplotlib._types.CapStyle). It looks a bit cluttered:

- The supported values blend into the bulk of text and are hard to find.
- The demo image should be below the list of values.
- There's a lot of redundancy: Available values are mentioned 4 times: list including descriptions, in the example plot, in `input_description`, as attributes of the Enum.
- While I see the use-case `demo()` still feels a bit out of place. You can keep it for now, but please give it a docstring.

brunobeltran replied:
I agree that the output looks cluttered. I've:

- edited the docstrings to emphasize supported values, by making it a section
- moved the image to the bottom
- removed all automatic member documentation from the enum to prevent any redundacy

timhoffm replied:
The above are good steps in the right direction, but I'm still somewhat worried that the values are difficult to find. They blend into the paragraphs of the bullet list.

Visibility could be further enhanced with a definition list. E.g.

~~~
    **Supported values:**

    .. rst-class:: value-list

       'miter'
          the "arrow-tip" style. Each boundary of the filled-in area
          will extend in a straight line parallel to the tangent vector of the
          centerline at the point it meets the corner, until they meet in a
          sharp point.
       'round'
          stokes every point within a radius of ``linewidth/2`` of the
          center lines.
       'bevel'
          the "squared-off" style. It can be thought of as a rounded
          corner where the "circular" part of the corner has been cut off.
~~~

Will result in `<dl class="value-list">...</dl>` so that we can CSS-style based on `.value-list`.

This is the lightweight version of making your own `.. value-list::` role, which may or may not be reasonable later on.

jklymak commented:
I guess I'm less conservative, (and its probably bothered a number of users who had edge cases, but)....

Ignoring JoinStyle and CapStyle for now, which I'm sure 99.999% of users have no idea exist: 

In the case of linestyle, a downstream library is going to do `style=line.get_linestyle()`.  What are they possibly going to do with the `style` variable?   I would guess it would be vanishingly rare that they are parsing the string?  Anyone who is, would probably welcome a change that allows them to *stop* parsing the string and instead manipulate a class object that had the proper levers...  Of course we could always provide a string `__repr__`...

But probably my imagination has failed me and there are other uses for the string?

timhoffm commented on lib/matplotlib/cbook/deprecation.py:
```diff
@@ -4,6 +4,41 @@
 
 
+# has to be here to prevent circular import, since it's used by `deprecated`
+class _classproperty:
```
This movement and the changes in `MatplotlibDeprecationWarning` seem unrelated. While needed here, they are too complex to be burried in a large change. I'd prefer them at least in a seprate commit, better in a dedicated PR. (My brain is not large enough to think this though while keeping all the style changes in mind).

anntzer replied:
I think you can just do a local import within the implementation of `cbook.deprecated`?

brunobeltran replied:
Yup, this should have been separate from the start. @anntzer I'll make that change in #18817 if that's the preferred way to fix this problem.

timhoffm commented on lib/matplotlib/lines.py:
```diff
@@ -1302,97 +1312,99 @@ def update_from(self, other):
-        s : {'miter', 'round', 'bevel'}
-            For examples see :doc:`/gallery/lines_bars_and_markers/joinstyle`.
+        s : `.JoinStyle`
+            Testing
```
What is this?

brunobeltran replied:
An out-of-hand vim macro, thanks for the catch!

timhoffm commented on lib/matplotlib/lines.py:
```diff
@@ -1302,97 +1312,99 @@ def update_from(self, other):
     def is_dashed(self):
         """
-        Return whether line has a dashed linestyle.
+        Return whether line has a dashed linestyle. A custom linestyle is
+        assumed to be dashed, we do not inspect the ``onoffseq`` directly.
```
```suggestion
        Return whether line has a dashed linestyle.
        
        A custom linestyle is assumed to be dashed, we do not inspect
        the ``onoffseq`` directly.
```
Summary line should be separated from following sentences.

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,179 @@
+"""
+Style description information that is shared across unrelated classses.
```
Since the module is `_types`. Shouldn't we somehow use "types" instead of "style description". "style" seems a bit too specific.

brunobeltran replied:
New docstring states the real goal of the class better, I hope.

timhoffm reviewed (commented):
It's making progress. :+1: 

My main concern is the layout of the rendered style docs. It's not as easily readable as it should be.

I'm also struggling with the `demo()` concept. In particular I bemoan the deletion of the examples. While I agree that they are currently cluttered into other kinds of examples, I had the idea that it would be nice to have an example section with all these reference examples (linestyles, marker, hatch, ...). The gallery pictures give a nicer index than the plain API-doc list. Of course, that's going against the idea that everything should be in one place. - For now, this is just a collection of thoughts. I'm not saying it has to be one way or the other.

timhoffm commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,179 @@
+        from matplotlib._types import CapStyle
+        CapStyle.demo()
+
+    Available options:
```
```suggestion
    Available values:
```
to match `CapStyle(value)` signature.

Not quite sure with "available". Alternatives: "Supported values:", "Values:"

brunobeltran commented:
> My main concern is the layout of the rendered style docs. It's not as easily readable as it should be.
Hopefully this is at least (partially) resolved. I look forward to hearing your take on the updates.
 
> In particular I bemoan the deletion of the examples. While I agree that they are currently cluttered into other kinds of examples, I had the idea that it would be nice to have an example section with all these reference examples (linestyles, marker, hatch, ...). 
I think this is totally in scope and attainable for my GSOD, and I don't think it's necessarily harder to do this if we delete the examples. Just because we don't use sphinx-gallery, doesn't mean it wouldn't be easy to do, since it would basically just involve linking the output of `_types.*.demo` to the appropriate class documentation for each type.

> The gallery pictures give a nicer index than the plain API-doc list. 
100% agree. I think giving the user the best view possible should take priority over any theoretical benefits from having everything "in one place". I think what matters is using `_types.py` to make sure nothing can go *out of sync*, even if it's not completely centralized.

timhoffm commented on doc/api/_types.rst:
```diff
@@ -0,0 +1,18 @@
+``matplotlib._types``
+**********************
+
+.. automodule:: matplotlib._types
```
Ok for now.

When we get more entries, every entry should probably get it's own sub-page. But since we might fiddle with the organization/representation anyway (c.f. sphinx-gallery), it's not worth to work on this right now.

timhoffm commented on doc/conf.py:
```diff
@@ -18,6 +18,8 @@
 import sphinx
 
 from datetime import datetime
+import warnings
+warnings.filterwarnings('ignore')
```
Is this intentional? I don't think we should globally disable warnings for doc builds.

brunobeltran replied:
That was an overly aggressive `git add -A`, thanks!

brunobeltran commented:
Okie dokie, @timhoffm, I rebased this now that the deprecation stuff has gone in, no real changes since your last review so I think it might be ready for a last look?

brunobeltran commented on doc/api/_types.rst:
```diff
@@ -0,0 +1,18 @@
+
+   .. autoclass:: JoinStyle
+      :members: demo
+      :exclude-members: bevel, miter, round, input_description
```
Yeah I checked, and you do need to exclude-members, for whatever reason. I'm not sure if this is a bug, as I also read the documentation to mean that "members" is an exclusive list, but phenomenologically this is what is needed.

timhoffm replied:
Strange, but :shrug:.

brunobeltran commented on doc/api/_types.rst:
```diff
@@ -0,0 +1,18 @@
+   :no-members:
+   :show-inheritance:
+
+   .. autodata:: _AutoStringNameEnum
```
So I've taken this out but @tacaswell had explicitly asked in a previous call that we make sure to include everything in the documentation, even if private, for purposes of completeness. I guess we just need to decide one way or the other?

timhoffm replied:
As a rule of thumb: I'd include everything that will be used (and seen) by others. That includes public API, private base classes of public classes, internal helpers which are used by developers throughout the library (e.g. the new `_api` module).

IMHO `_AutoStringNameEnum` doesn't qualify for this. It's too specialized to be of any help in the docs. Moreover, this part of the doc is end-user facing. And they certainly don't need to know about this helper.

brunobeltran commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,182 @@
+"""
+Types that would not normally have their own class are documented here.
```
I tried to do this better, let me know if you still need changes.

timhoffm replied:
> Concepts used by the matplotlib API that do not yet have a dedicated class.

is still confusing to me: "not yet" alludes to a future intended change without being clear. This leaves the reader with an impression that the below content is only provisional. Additionally, you *are* writing dedicated classes below?!?

Probably needs some more careful wording. The concepts are internally implemented by these types, but the types are not used in the public user interface.

jklymak commented on examples/lines_bars_and_markers/joinstyle.py:
```diff
@@ -1,90 +0,0 @@
-"""
-==========================
```
Why was this example deleted?  If you moved the docs to a class defn, fine, but leave the stub here with a link to the docs?  Maybe w/ a small minimal example?

brunobeltran replied:
I didn't feel like it made sense to have the exact same example in two different places. My goal with these API docs was to be a strictly-better drop-in replacement for this example. 

Is the concern here that the code that makes (what is now) the `JoinStyle.demo` figure is not as easily accessible in the docs anymore? I would propose that the example was not really adding much to the documentation via its code, but only really via its output (the visual representation of each Join/CapStyle is what's important).

Is the concern discoverability? I know that we have issues with Google sending people to old versions of the docs, but presumably the hope is that the new page quickly gets indexed as the go-to place for JoinStyle. In which case the relevant example figure will still be there.

If the concern is that people looking for that specific example will be confused as to where it went, then my impression of what's the best way to leave a "stub" behind is, instead of leaving an actual stub page with a link, we should really just use something like [the reredirects package](https://pypi.org/project/sphinx-reredirects/) to simply have an HTTP 301 response whenever someone tries to visit that page. As a side-effect this will also help with Google search indexing.

jklymak replied:
Something like the redirects package looks very interesting, though version 0.0.0 is a negative.

But the other concern is that the example will no longer will show up in the example gallery...

brunobeltran replied:
Haha fair, didn't realize the version number was so sketch, but frankly I can white room that package in a couple hours if we wanted to have our own implementation.

Gitlab reports that the whole "package" is only 62 lines of Python code. Was mostly linking it as a proof-of-principle that this is not impossible to do inside of Sphinx.

This may be totally off-base, but I struggle to imagine the user that scrolls through the examples gallery and is enticed to click on the JoinStyle demo in particular. Unless you know what it is already? But people that already know the name for something typically find examples by getting directly linked in via Google...

If I'm outvoted on that point, I guess we would just add two stub examples, one which just calls `CapStyle.demo` and one that calls `JoinStyle.demo`?

jklymak replied:
If you are browsing the examples, and come across it you learn something new.  If you don't know what a join style is, you can never google it.  People need *some* way to know about the features in the library, and the example pages are basically it.  

If you just want to have it call and link the class demo, thats perhaps fine.  Then the plot shows up, and the user is linked to complete docs.  But I think leaving it in as an example in the gallery is useful.

QuLogic replied:
I think adding redirects is not the issue, but collating through the hundreds of old versions is a chore.

As for examples, I thought there was some talk of putting a smaller set of examples for all plot types, and maybe we could do the same for all these new style classes?

jklymak replied:
^ that would be great.  However I don't think we should delete existing examples until that is created?

timhoffm replied:
:+1: for leaving stub examples and linking to the class. and later collecting these examples in a dedicated section.

(This is for discoverability rather than for keeping links alive).

jklymak commented on lib/matplotlib/backend_bases.py:
```diff
@@ -820,10 +820,8 @@ def get_antialiased(self):
-        """
-        return self._capstyle
+        """Return the `.CapStyle`."""
+        return self._capstyle.name
```
We have discussed this before, but my vote is that the minuscule breakage here of just returning `self._capstyle` is worth it.  I can't imagine anyone programatically wanting to call this...

However, you *may* want to implement an `__equals__` operator for these classes so `capstyle == 'round'` returns True.  I'm not clear from your mixing above whether that is already the case.  If so, then I really do not support keeping the type the same, since the odds of someone checking the type here are miniscule.

brunobeltran replied:
This is a super-good point, and I even left in the commit in this PR (reverted by a later commit) which accomplishes this, in case we ever change our minds.

Since we use `str` as a mix-in, we get `__eq__`, `__add__`, etc, etc all for free from `str`. The long term goal is for by the end of GSOD (around March), I want to have *all* of the "styles" encapsulated in this same kind of new-style class (linestyle, hatch, joinstyle, capstyle, color, etc.). The original reason why this change was veto'd is that it seemed like a bad idea to introduce these new objects into our API if only one or two of them are complete (currently, only joinstyle/capstyle, although linestyle/hatch are *very* close to done).

The current plan (if I understand the latest state of the discussion correctly) is to put in all the classes, and then sometime around the end of GSOD when they're all started to "get in", make a separate PR that does all the API breaking at once, preferably with some loud announcements?

As I've brought up in the GSOD call, my initial impression is that this doesn't really make sense for something like `JoinStyle`, where the only API break is going to be if someone asks `type(js)`. However, @tacaswell is right to be dubious of this approach, as I do have to admit that even when very accomplished, careful developers migrate something from a string to a more advanced class, there tends to be very subtle ways in which this breaks stuff for users (I'm thinking for example of Python's move to `dict_keys` objects).

brunobeltran replied:
Almost forgot. In the current state of the PR, there is one *real life* use case then in which the new class object can't be used directly in place of the old strings. Since `Enum` implements `__hash__`, then if a backend is using the `JoinStyle` to query a dictionary like:
```python
# taken from SVG backend verbatim
_capstyle_d = {'projecting': 'square', 'butt': 'butt', 'round': 'round'}
```
Then the following raises a `KeyError`. 
```python
_capstyle_d[CapStyle('projecting')]
```

This means that likely all custom backends that handle join or capstyles would be broken by this PR if we let it go in as-is.

However, this is easy enough to fix:
```python
# in the superclass
def __hash__(self):
    return str(self).__hash__()
```
will make `_capstyle_d[CapStyle('projecting')]` work again.

anntzer commented on lib/matplotlib/_types.py:
```diff
@@ -0,0 +1,204 @@
+class _AutoStringNameEnum(Enum):
+    """Automate the ``name = 'name'`` part of making a (str, Enum)."""
+    def _generate_next_value_(name, start, count, last_values):
+        return name
```
May be worth noting that this will become unneeded (it is the default behavior) in Py3.10 per https://bugs.python.org/issue42385.

brunobeltran commented:
Alright, fixed @timhoffm's last remaining wording complaint, put gallery entries back in, and added hash equivalence, to make it easy to expose the new classes in a follow-up PR.

This might be ready to go in now?

timhoffm commented:
@brunobeltran test failure is real.

brunobeltran commented:
Oops, overly greedy `sed` call, checked it more carefully this time.

timhoffm reviewed (approved):
Looks good now. Only needs a final rebase. Sorry for the delayed review.

timhoffm commented:
Ping @brunobeltran. The (modified) plot directive stuff is merged, so you should rebase on that.

brunobeltran commented:
Rebased, so should be good post-CI!

timhoffm commented:
Sorry, something seems to have changed again.
